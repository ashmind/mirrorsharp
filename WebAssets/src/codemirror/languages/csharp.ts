import { LRLanguage } from '@codemirror/language';
import { styleTags, tags } from '@lezer/highlight';
import { parser } from 'lezer-csharp-simple';

const csharpSyntax = LRLanguage.define({
    parser: parser.configure({
        props: [styleTags({
            Keyword: tags.keyword,
            Comment: tags.comment,
            Number: tags.number,
            String: tags.string,
            Punctuation: tags.punctuation
        })]
    }),
    languageData: {
        closeBrackets: { brackets: ['(', '[', '{', "'", '"', '`'] },
        commentTokens: { line: '//', block: { open: '/*', close: '*/' } }
    }
});

export const csharp = csharpSyntax.extension;