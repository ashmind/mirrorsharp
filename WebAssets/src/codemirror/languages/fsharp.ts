import { StreamLanguage } from '@codemirror/language';
import { fSharp } from '@codemirror/legacy-modes/mode/mllike';

export const fsharp = StreamLanguage.define(fSharp).extension;