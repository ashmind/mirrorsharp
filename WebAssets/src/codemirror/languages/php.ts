import { phpLanguage } from '@codemirror/lang-php';
import { styleTags, tags } from '@lezer/highlight';

export const php = phpLanguage.configure({
    props: [styleTags({
        Keyword: tags.keyword,
        Comment: tags.comment,
        Number: tags.number,
        String: tags.string,
        Punctuation: tags.punctuation
    })]
}).extension;