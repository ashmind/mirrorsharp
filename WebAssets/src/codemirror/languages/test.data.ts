const normalize = (code: string) => code.replace(/\r\n|\r|\n/g, '\r\n').trim();

export const CODE_CSHARP = normalize(`
public class C<T> {
    // test comment
    public void M<U>() {
        double d = 1.2e3;
        string s1 = "test";
        string s2 = $"a{s1}b";
        char c = 't';
        Action a = () => {};
    }
}`);

export const CODE_VB = normalize(`
Public Class C (Of T)
    ' test comment
    Public Sub M (Of U)()
        Dim d As Decimal = 1.2e3
        Dim s1 As String = "test"
        Dim s2 As String = $"a{s1}b"
        Dim c As Char = 't'
        Dim a As Action = Sub()
                          End Sub
    End Sub
End Class
`);

export const CODE_FSHARP = normalize(`
type C<'a> =
    // test comment
    member this.M<'b>() =
        let d = 1.2e3
        let s1 = "test"
        let s2 = $"a{s1}b"
        let c = 't'
        let a = fun () -> ()
        ()
`);

export const CODE_IL = normalize(`
.class public C\`1<T>
    extends [System.Runtime]System.Object
{
    // test comment
    .method public
        instance void M<U> () cil managed
    {
        .maxstack 3
        .locals init (
            [0] float64 d,
            [1] string s1
        )

        IL_0000: ldc.r8 1.2e3
        IL_0009: stloc.0
        IL_000a: ldstr "test"
        IL_000f: stloc.1
        IL_0010: ret
    }
}
`);

export const CODE_PHP = normalize(`
<?php

class C {
    // test comment
    public function M() {
        $d = 1.2e3;
        $s1 = 'test';
        $s2 = "a{$s1}b";
        $a = fn() => 0;
    }
}
`);