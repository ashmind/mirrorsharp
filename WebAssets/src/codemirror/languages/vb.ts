import { StreamLanguage } from '@codemirror/language';
import { vb as vbImport } from '@codemirror/legacy-modes/mode/vb';

// Test-only:
// ts-unused-exports:disable-next-line
export const vbLanguage = StreamLanguage.define(vbImport);

export const vb = vbLanguage.extension;