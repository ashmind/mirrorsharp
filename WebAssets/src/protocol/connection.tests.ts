import { installMockSocket, MockSocket } from '../testing/shared/mock-socket';
import { TestDriver } from '../testing/test-driver-jest';
import { Connection } from './connection';

test('attempts to reopen socket if closed', async () => {
    const driver = await TestDriver.new({});

    driver.socket.close();
    await driver.completeBackgroundWork();

    expect(driver.socket.createdCount).toBe(2);
});

test('does not attempt to reopen socket if explicitly closed', async () => {
    const socket = installMockSocket(new MockSocket()).mock;
    const connection = new Connection('_', { closed: false });

    connection.close();
    socket.close();

    jest.advanceTimersByTime(1000);

    expect(socket.createdCount).toBe(1);
});

test('attempts to reopen socket at expected intervals', async () => {
    const driver = await TestDriver.new({});

    const retryIntervals = [] as Array<number>;
    let last = {
        createdCount: driver.socket.createdCount,
        retryTime: 0
    };

    driver.socket.close();
    let time = 0;
    for (; time < 180; time += 1) {
        jest.advanceTimersByTime(1000);
        const { createdCount } = driver.socket;
        if (createdCount > last.createdCount) {
            retryIntervals.push(time - last.retryTime);
            last = { createdCount, retryTime: time };
        }
        if (driver.socket.readyState === MockSocket.OPEN)
            driver.socket.close();
    }

    expect(retryIntervals).toEqual([0, 1, 2, 4, 8, 16, 32, 60]);
});