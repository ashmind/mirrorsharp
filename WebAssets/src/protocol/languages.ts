export const LANGUAGE_CSHARP = 'C#';
export const LANGUAGE_VB = 'Visual Basic';
export const LANGUAGE_FSHARP = 'F#';
export const LANGUAGE_PHP = 'PHP';
export const LANGUAGE_IL = 'IL';

export const LANGUAGE_DEFAULT = LANGUAGE_CSHARP;

export type Language = typeof LANGUAGE_CSHARP
    | typeof LANGUAGE_VB
    | typeof LANGUAGE_FSHARP
    | typeof LANGUAGE_PHP
    | typeof LANGUAGE_IL;