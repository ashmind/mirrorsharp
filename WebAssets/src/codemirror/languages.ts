import { LANGUAGE_CSHARP, LANGUAGE_FSHARP, LANGUAGE_IL, LANGUAGE_PHP, LANGUAGE_VB } from '../protocol/languages';
import { cil } from './languages/cil';
import { csharp } from './languages/csharp';
import { fsharp } from './languages/fsharp';
import { php } from './languages/php';
import { vb } from './languages/vb';

export const languageExtensions = {
    [LANGUAGE_CSHARP]: csharp,
    [LANGUAGE_VB]: vb,
    [LANGUAGE_FSHARP]: fsharp,
    [LANGUAGE_PHP]: php,
    [LANGUAGE_IL]: cil
} as const;