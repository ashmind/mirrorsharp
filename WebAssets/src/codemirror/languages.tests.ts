import { LANGUAGE_FSHARP, LANGUAGE_IL, LANGUAGE_PHP, LANGUAGE_VB } from '../protocol/languages';
import { TestDriver } from '../testing/test-driver-jest';
import { CODE_CSHARP, CODE_FSHARP, CODE_IL, CODE_PHP, CODE_VB } from './languages/test.data';

test('C# highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({ text: CODE_CSHARP });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});

test('VB highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({
        language: LANGUAGE_VB,
        text: CODE_VB
    });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});

test('F# highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({
        language: LANGUAGE_FSHARP,
        text: CODE_FSHARP
    });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});

test('IL highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({
        language: LANGUAGE_IL,
        text: CODE_IL
    });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});

test('PHP highlighting applies expected classes', async () => {
    const driver = await TestDriver.new({
        language: LANGUAGE_PHP,
        text: CODE_PHP
    });
    await driver.completeBackgroundWork();

    const html = driver.getCodeMirrorView().contentDOM.innerHTML;

    expect(html).toMatchSnapshot();
});