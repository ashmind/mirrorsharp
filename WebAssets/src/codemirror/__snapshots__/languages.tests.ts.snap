// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`C# highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-keyword">public</span> <span class="tok-keyword">class</span> C<span class="tok-punctuation">&lt;</span>T<span class="tok-punctuation">&gt;</span> <span class="tok-punctuation">{</span></div><div class="cm-line">    <span class="tok-comment">// test comment</span></div><div class="cm-line">    <span class="tok-keyword">public</span> <span class="tok-keyword">void</span> M<span class="tok-punctuation">&lt;</span>U<span class="tok-punctuation">&gt;</span><span class="tok-punctuation">(</span><span class="tok-punctuation">)</span> <span class="tok-punctuation">{</span></div><div class="cm-line">        <span class="tok-keyword">double</span> d <span class="tok-punctuation">=</span> <span class="tok-number">1.2e3</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-keyword">string</span> s1 <span class="tok-punctuation">=</span> <span class="tok-string">"test"</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-keyword">string</span> s2 <span class="tok-punctuation">=</span> $<span class="tok-string">"a{s1}b"</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-keyword">char</span> c <span class="tok-punctuation">=</span> 't'<span class="tok-punctuation">;</span></div><div class="cm-line">        Action a <span class="tok-punctuation">=</span> <span class="tok-punctuation">(</span><span class="tok-punctuation">)</span> <span class="tok-punctuation">=&gt;</span> <span class="tok-punctuation">{</span><span class="tok-punctuation">}</span><span class="tok-punctuation">;</span></div><div class="cm-line">    <span class="tok-punctuation">}</span></div><div class="cm-line"><span class="tok-punctuation">}</span></div>"`;

exports[`F# highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-keyword">type</span> <span class="tok-variableName">C</span><span class="tok-operator">&lt;</span>'<span class="tok-variableName">a</span><span class="tok-operator">&gt;</span> <span class="tok-operator">=</span></div><div class="cm-line">    <span class="tok-comment">// test comment</span></div><div class="cm-line">    <span class="tok-keyword">member</span> <span class="tok-variableName">this</span><span class="tok-operator">.</span><span class="tok-variableName">M</span><span class="tok-operator">&lt;</span>'<span class="tok-variableName">b</span><span class="tok-operator">&gt;</span>() <span class="tok-operator">=</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">d</span> <span class="tok-operator">=</span> <span class="tok-number">1.2e3</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">s1</span> <span class="tok-operator">=</span> <span class="tok-string">"test"</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">s2</span> <span class="tok-operator">=</span> $<span class="tok-string">"a{s1}b"</span></div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">c</span> <span class="tok-operator">=</span> '<span class="tok-variableName">t</span>'</div><div class="cm-line">        <span class="tok-keyword">let</span> <span class="tok-variableName">a</span> <span class="tok-operator">=</span> <span class="tok-keyword">fun</span> () <span class="tok-operator">-</span><span class="tok-operator">&gt;</span> ()</div><div class="cm-line">        ()</div>"`;

exports[`IL highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-keyword">.class</span> <span class="tok-keyword">public</span> C\`1&lt;T&gt;</div><div class="cm-line">    <span class="tok-keyword">extends</span> [System.Runtime]System.Object</div><div class="cm-line">{</div><div class="cm-line">    <span class="tok-comment">// test comment</span></div><div class="cm-line">    <span class="tok-keyword">.method</span> <span class="tok-keyword">public</span></div><div class="cm-line">        <span class="tok-keyword">instance</span> void M&lt;U&gt; () <span class="tok-keyword">cil</span> <span class="tok-keyword">managed</span></div><div class="cm-line">    {</div><div class="cm-line">        <span class="tok-keyword">.maxstack</span> <span class="tok-number">3</span></div><div class="cm-line">        <span class="tok-keyword">.locals</span> <span class="tok-keyword">init</span> (</div><div class="cm-line">            [0] float64 d,</div><div class="cm-line">            [1] string s1</div><div class="cm-line">        )</div><div class="cm-line"><br></div><div class="cm-line">        <span class="tok-typeName">IL_0000:</span> <span class="tok-variableName">ldc.r8</span> <span class="tok-number">1</span>.2e3</div><div class="cm-line">        <span class="tok-typeName">IL_0009:</span> <span class="tok-variableName">stloc.0</span></div><div class="cm-line">        <span class="tok-typeName">IL_000a:</span> <span class="tok-variableName">ldstr</span> <span class="tok-string">"test"</span></div><div class="cm-line">        <span class="tok-typeName">IL_000f:</span> <span class="tok-variableName">stloc.1</span></div><div class="cm-line">        <span class="tok-typeName">IL_0010:</span> <span class="tok-variableName">ret</span></div><div class="cm-line">    }</div><div class="cm-line">}</div>"`;

exports[`PHP highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-meta">&lt;?php</span></div><div class="cm-line"><br></div><div class="cm-line"><span class="tok-keyword">class</span> <span class="tok-className">C</span> <span class="tok-punctuation">{</span></div><div class="cm-line">    <span class="tok-comment">// test comment</span></div><div class="cm-line">    <span class="tok-keyword">public</span> <span class="tok-keyword">function</span> <span class="tok-variableName">M</span><span class="tok-punctuation">(</span><span class="tok-punctuation">)</span> <span class="tok-punctuation">{</span></div><div class="cm-line">        <span class="tok-variableName">$d</span> <span class="tok-operator">=</span> <span class="tok-number">1.2e3</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-variableName">$s1</span> <span class="tok-operator">=</span> <span class="tok-string">'test'</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-variableName">$s2</span> <span class="tok-operator">=</span> <span class="tok-string">"a</span><span class="tok-punctuation">{</span><span class="tok-variableName">$s1</span><span class="tok-punctuation">}</span><span class="tok-string">b"</span><span class="tok-punctuation">;</span></div><div class="cm-line">        <span class="tok-variableName">$a</span> <span class="tok-operator">=</span> <span class="tok-keyword">fn</span><span class="tok-punctuation">(</span><span class="tok-punctuation">)</span> <span class="tok-punctuation">=&gt;</span> <span class="tok-number">0</span><span class="tok-punctuation">;</span></div><div class="cm-line">    <span class="tok-punctuation">}</span></div><div class="cm-line"><span class="tok-punctuation">}</span></div>"`;

exports[`VB highlighting applies expected classes 1`] = `"<div class="cm-line"><span class="tok-keyword">Public</span> <span class="tok-keyword">Class</span> <span class="tok-variableName">C</span> (<span class="tok-keyword">Of</span> <span class="tok-variableName">T</span>)</div><div class="cm-line">    <span class="tok-comment">' test comment</span></div><div class="cm-line">    <span class="tok-keyword">Public</span> <span class="tok-keyword">Sub</span> <span class="tok-variableName">M</span> (<span class="tok-keyword">Of</span> <span class="tok-variableName">U</span>)()</div><div class="cm-line">        <span class="tok-keyword">Dim</span> <span class="tok-variableName">d</span> <span class="tok-keyword">As</span> <span class="tok-keyword">Decimal</span> = <span class="tok-number">1.2</span><span class="tok-variableName">e3</span></div><div class="cm-line">        <span class="tok-keyword">Dim</span> <span class="tok-variableName">s1</span> <span class="tok-keyword">As</span> <span class="tok-keyword">String</span> = <span class="tok-string">"test"</span></div><div class="cm-line">        <span class="tok-keyword">Dim</span> <span class="tok-variableName">s2</span> <span class="tok-keyword">As</span> <span class="tok-keyword">String</span> = <span class="tok-invalid">$</span><span class="tok-string">"a{s1}b"</span></div><div class="cm-line">        <span class="tok-keyword">Dim</span> <span class="tok-variableName">c</span> <span class="tok-keyword">As</span> <span class="tok-keyword">Char</span> = <span class="tok-comment">'t'</span></div><div class="cm-line">        <span class="tok-keyword">Dim</span> <span class="tok-variableName">a</span> <span class="tok-keyword">As</span> <span class="tok-variableName">Action</span> = <span class="tok-keyword">Sub</span>()</div><div class="cm-line">                          <span class="tok-keyword">End</span> <span class="tok-keyword">Sub</span></div><div class="cm-line">    <span class="tok-keyword">End</span> <span class="tok-keyword">Sub</span></div><div class="cm-line"><span class="tok-keyword">End</span> <span class="tok-keyword">Class</span></div>"`;
